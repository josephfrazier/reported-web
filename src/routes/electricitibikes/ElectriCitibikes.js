/**
 * React Starter Kit (https://www.reactstarterkit.com/)
 *
 * Copyright Â© 2014-present Kriasoft, LLC. All rights reserved.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE.txt file in the root directory of this source tree.
 */

import React from 'react';
import PropTypes from 'prop-types';
import promisedLocation from 'promised-location';
import humanizeDistance from 'humanize-distance';
import geodist from 'geodist';
import 'intl/locale-data/jsonp/en.js'; // https://github.com/andyearnshaw/Intl.js/issues/271#issuecomment-292233493
import strftime from 'strftime';
import PolygonLookup from 'polygon-lookup';
import geolib from 'geolib';

import withStyles from 'isomorphic-style-loader/lib/withStyles';
import s from './ElectriCitibikes.css';

if (!global.window) {
  global.window = require('global/window'); // eslint-disable-line global-require
}
const streamdataio = require('streamdataio-js-sdk'); // eslint-disable-line global-require
const jsonpatch = require('fast-json-patch'); // eslint-disable-line global-require

class ElectriCitibikes extends React.Component {
  static propTypes = {
    title: PropTypes.string.isRequired,
  };

  state = {
    data: {
      features: [],
    },
  };

  componentDidMount() {
    const appToken = 'OTY5YTM3ZmItNTA2Ni00ZThhLWJmNzItYjVmM2QwYzZlMmYy';

    const eventSource = streamdataio.createEventSource(
      'https://bikeangels-api.citibikenyc.com/map/v1/nyc/stations',
      appToken,
    );

    eventSource
      .onOpen(() => {
        console.info('streamdata Event Source connected.');
      })
      .onData(data => {
        console.info({ data });
        this.updateData({ data });
      })
      .onPatch(patch => {
        console.info({ patch });

        const data = jsonpatch.deepClone(this.state.data);
        jsonpatch.applyPatch(data, patch);
        this.updateData({ data });

        const changedEbikes = patch.filter(operation =>
          operation.path.includes('ebikes_available'),
        );
        changedEbikes.forEach(operation => {
          const { op, path, value } = operation;
          const changedStation = data.features[path.split('/')[2]].properties;
          const ebikeCount = op === 'remove' ? 0 : value;
          console.info(`${ebikeCount} ebikes at ${changedStation.name}`);
        });
      });

    // open the data stream to the REST service through streamdata.io proxy
    // eslint-disable-next-line no-underscore-dangle
    eventSource.open()._sse.addErrorListener(() => {
      window.location.reload();
    });

    fetch(
      'https://services5.arcgis.com/GfwWNkhOj9bNBqoJ/arcgis/rest/services/nybb/FeatureServer/0/query?where=1=1&outFields=*&outSR=4326&f=geojson',
    )
      .then(response => response.json())
      .then(boroughBoundariesFeatureCollection => {
        this.setState({
          boroughBoundariesFeatureCollection,
        });
      });
  }

  async updateData({ data }) {
    const updatedAt = Date.now();
    this.setState({ data, updatedAt });

    promisedLocation().then(({ coords: { latitude, longitude } }) =>
      this.setState({ latitude, longitude }),
    );
  }

  render() {
    const {
      state: {
        data,
        latitude,
        longitude,
        updatedAt,
        boroughBoundariesFeatureCollection,
      },
    } = this;
    return (
      <div className={s.root}>
        <div className={s.container}>
          <h1>{this.props.title}</h1>
          <ElectriCitibikeList
            data={data}
            latitude={latitude}
            longitude={longitude}
            updatedAt={updatedAt}
            boroughBoundariesFeatureCollection={
              boroughBoundariesFeatureCollection
            }
          />
        </div>
      </div>
    );
  }
}

function getMapUrl({ station }) {
  return `https://www.google.com/maps?hl=en&q=${station.latitude},${
    station.longitude
  }`;
}

export function ElectriCitibikeList({
  data,
  latitude,
  longitude,
  updatedAt,
  boroughBoundariesFeatureCollection,
}) {
  const lookup = new PolygonLookup(boroughBoundariesFeatureCollection);

  const stations = data.features.map(f => {
    const { coordinates } = f.geometry;
    const start = { latitude, longitude };
    const end = {
      latitude: coordinates[1],
      longitude: coordinates[0],
    };
    let dist = 'unknown distance';
    let distMeters;
    let compassBearing;
    if (start.latitude && start.longitude) {
      dist = humanizeDistance(start, end, 'en-US', 'us');
      distMeters = geodist(start, end, { unit: 'meters', exact: true });
      compassBearing = geolib.getCompassDirection(start, end).exact;
    }

    const boroughPolygon = lookup.search(end.longitude, end.latitude);

    return {
      ...f.properties,
      latitude: end.latitude,
      longitude: end.longitude,
      dist,
      distMeters,
      boroughPolygon,
      compassBearing,
    };
  });

  const ebikeStations = stations.filter(
    station => station.ebikes_available > 0,
  );
  ebikeStations.sort((a, b) => a.distMeters - b.distMeters);
  const humanDate = strftime('%r', new Date(updatedAt));
  const totalEbikesAvailable = ebikeStations
    .map(station => station.ebikes_available)
    .reduce((a, b) => a + b, 0);
  return (
    <>
      {totalEbikesAvailable} available as of {humanDate}
      {ebikeStations.map(station => (
        <details key={station.name} style={{ margin: '1rem 0' }}>
          <summary>
            {station.ebikes_available} @&nbsp;
            <a
              target="_blank"
              rel="noopener noreferrer"
              href={getMapUrl({ station })}
            >
              {station.name}, {station.boroughPolygon.properties.BoroName}
            </a>
            <br />
            ({station.dist} {station.compassBearing} from you, about{' '}
            {Math.ceil(station.distMeters / 80)} blocks)
          </summary>
          <ul>
            {station.ebikes &&
              station.ebikes.map(ebike => (
                <li key={ebike.bike_number}>
                  {`#${ebike.bike_number}`} has {ebike.charge}/4 charge
                </li>
              ))}
          </ul>
        </details>
      ))}
    </>
  );
}

export default withStyles(s)(ElectriCitibikes);
