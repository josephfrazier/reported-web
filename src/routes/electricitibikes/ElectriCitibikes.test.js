/* eslint-env jest */
/* eslint-disable padded-blocks, no-unused-expressions */

import '@babel/polyfill';
import React from 'react';
import renderer from 'react-test-renderer';
import { ElectriCitibikeList } from './ElectriCitibikes.js';

import {
  data,
  latitude,
  longitude,
  updatedAt,
  boroughBoundariesFeatureCollection,
} from './test_state.json';

require('timezone-mock').register('US/Eastern');

describe('ElectriCitibikeList', () => {
  test('renders children correctly', () => {
    const wrapper = renderer
      .create(
        <ElectriCitibikeList
          data={data}
          latitude={latitude}
          longitude={longitude}
          updatedAt={updatedAt}
          boroughBoundariesFeatureCollection={
            boroughBoundariesFeatureCollection
          }
        />,
      )
      .toJSON();

    expect(wrapper).toMatchSnapshot();
  });
});
