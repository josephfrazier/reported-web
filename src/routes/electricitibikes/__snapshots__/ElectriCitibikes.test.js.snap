// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ElectriCitibikeList renders children correctly 1`] = `
Array [
  "15",
  " available as of ",
  "07:50:57 PM",
  <details
    style={
      Object {
        "margin": "1rem 0",
      }
    }
  >
    <summary>
      1
       @ 
      <a
        href="https://www.google.com/maps?hl=en&q=40.7128,-74.006"
        rel="noopener noreferrer"
        target="_blank"
      >
        Bank St & Washington St
      </a>
      <br />
      (
      1.6 mi
       away, about 
      33
       
      blocks)
    </summary>
    <ul />
  </details>,
  <details
    style={
      Object {
        "margin": "1rem 0",
      }
    }
  >
    <summary>
      1
       @ 
      <a
        href="https://www.google.com/maps?hl=en&q=40.7128,-74.006"
        rel="noopener noreferrer"
        target="_blank"
      >
        DeKalb Ave & Hudson Ave
      </a>
      <br />
      (
      2.1 mi
       away, about 
      42
       
      blocks)
    </summary>
    <ul>
      <li>
        #70027
         has 
        3
        /4 charge
      </li>
    </ul>
  </details>,
  <details
    style={
      Object {
        "margin": "1rem 0",
      }
    }
  >
    <summary>
      1
       @ 
      <a
        href="https://www.google.com/maps?hl=en&q=40.7128,-74.006"
        rel="noopener noreferrer"
        target="_blank"
      >
        Carlton Ave & Park Ave
      </a>
      <br />
      (
      2.1 mi
       away, about 
      42
       
      blocks)
    </summary>
    <ul>
      <li>
        #70021
         has 
        3
        /4 charge
      </li>
    </ul>
  </details>,
  <details
    style={
      Object {
        "margin": "1rem 0",
      }
    }
  >
    <summary>
      1
       @ 
      <a
        href="https://www.google.com/maps?hl=en&q=40.7128,-74.006"
        rel="noopener noreferrer"
        target="_blank"
      >
        Clermont Ave & Park Ave
      </a>
      <br />
      (
      2.2 mi
       away, about 
      44
       
      blocks)
    </summary>
    <ul>
      <li>
        #65015
         has 
        3
        /4 charge
      </li>
    </ul>
  </details>,
  <details
    style={
      Object {
        "margin": "1rem 0",
      }
    }
  >
    <summary>
      1
       @ 
      <a
        href="https://www.google.com/maps?hl=en&q=40.7128,-74.006"
        rel="noopener noreferrer"
        target="_blank"
      >
        8 Ave & W 31 St
      </a>
      <br />
      (
      2.7 mi
       away, about 
      54
       
      blocks)
    </summary>
    <ul>
      <li>
        #70081
         has 
        4
        /4 charge
      </li>
    </ul>
  </details>,
  <details
    style={
      Object {
        "margin": "1rem 0",
      }
    }
  >
    <summary>
      1
       @ 
      <a
        href="https://www.google.com/maps?hl=en&q=40.7128,-74.006"
        rel="noopener noreferrer"
        target="_blank"
      >
        Meserole Ave & Manhattan Ave
      </a>
      <br />
      (
      2.9 mi
       away, about 
      60
       
      blocks)
    </summary>
    <ul>
      <li>
        #51037
         has 
        1
        /4 charge
      </li>
    </ul>
  </details>,
  <details
    style={
      Object {
        "margin": "1rem 0",
      }
    }
  >
    <summary>
      1
       @ 
      <a
        href="https://www.google.com/maps?hl=en&q=40.7128,-74.006"
        rel="noopener noreferrer"
        target="_blank"
      >
        Bayard St & Leonard St
      </a>
      <br />
      (
      3 mi
       away, about 
      61
       
      blocks)
    </summary>
    <ul>
      <li>
        #70084
         has 
        3
        /4 charge
      </li>
    </ul>
  </details>,
  <details
    style={
      Object {
        "margin": "1rem 0",
      }
    }
  >
    <summary>
      1
       @ 
      <a
        href="https://www.google.com/maps?hl=en&q=40.7128,-74.006"
        rel="noopener noreferrer"
        target="_blank"
      >
        Throop Ave & Myrtle Ave
      </a>
      <br />
      (
      3.5 mi
       away, about 
      70
       
      blocks)
    </summary>
    <ul>
      <li>
        #70127
         has 
        3
        /4 charge
      </li>
    </ul>
  </details>,
  <details
    style={
      Object {
        "margin": "1rem 0",
      }
    }
  >
    <summary>
      1
       @ 
      <a
        href="https://www.google.com/maps?hl=en&q=40.7128,-74.006"
        rel="noopener noreferrer"
        target="_blank"
      >
        14 St & 7 Ave
      </a>
      <br />
      (
      3.6 mi
       away, about 
      72
       
      blocks)
    </summary>
    <ul />
  </details>,
  <details
    style={
      Object {
        "margin": "1rem 0",
      }
    }
  >
    <summary>
      1
       @ 
      <a
        href="https://www.google.com/maps?hl=en&q=40.7128,-74.006"
        rel="noopener noreferrer"
        target="_blank"
      >
        Broadway & W 60 St
      </a>
      <br />
      (
      4.1 mi
       away, about 
      83
       
      blocks)
    </summary>
    <ul>
      <li>
        #65004
         has 
        1
        /4 charge
      </li>
    </ul>
  </details>,
  <details
    style={
      Object {
        "margin": "1rem 0",
      }
    }
  >
    <summary>
      1
       @ 
      <a
        href="https://www.google.com/maps?hl=en&q=40.7128,-74.006"
        rel="noopener noreferrer"
        target="_blank"
      >
        E 74 St & 1 Ave
      </a>
      <br />
      (
      4.7 mi
       away, about 
      95
       
      blocks)
    </summary>
    <ul>
      <li>
        #70071
         has 
        3
        /4 charge
      </li>
    </ul>
  </details>,
  <details
    style={
      Object {
        "margin": "1rem 0",
      }
    }
  >
    <summary>
      1
       @ 
      <a
        href="https://www.google.com/maps?hl=en&q=40.7128,-74.006"
        rel="noopener noreferrer"
        target="_blank"
      >
        Riverside Dr & W 72 St
      </a>
      <br />
      (
      4.8 mi
       away, about 
      97
       
      blocks)
    </summary>
    <ul />
  </details>,
  <details
    style={
      Object {
        "margin": "1rem 0",
      }
    }
  >
    <summary>
      1
       @ 
      <a
        href="https://www.google.com/maps?hl=en&q=40.7128,-74.006"
        rel="noopener noreferrer"
        target="_blank"
      >
        W 84 St & Columbus Ave
      </a>
      <br />
      (
      5.3 mi
       away, about 
      107
       
      blocks)
    </summary>
    <ul>
      <li>
        #70208
         has 
        1
        /4 charge
      </li>
    </ul>
  </details>,
  <details
    style={
      Object {
        "margin": "1rem 0",
      }
    }
  >
    <summary>
      1
       @ 
      <a
        href="https://www.google.com/maps?hl=en&q=40.7128,-74.006"
        rel="noopener noreferrer"
        target="_blank"
      >
        W 88 St & West End Ave
      </a>
      <br />
      (
      5.6 mi
       away, about 
      112
       
      blocks)
    </summary>
    <ul>
      <li>
        #65020
         has 
        1
        /4 charge
      </li>
    </ul>
  </details>,
  <details
    style={
      Object {
        "margin": "1rem 0",
      }
    }
  >
    <summary>
      1
       @ 
      <a
        href="https://www.google.com/maps?hl=en&q=40.7128,-74.006"
        rel="noopener noreferrer"
        target="_blank"
      >
        W 95 St & Broadway
      </a>
      <br />
      (
      5.9 mi
       away, about 
      119
       
      blocks)
    </summary>
    <ul>
      <li>
        #51012
         has 
        1
        /4 charge
      </li>
    </ul>
  </details>,
]
`;
